lib/kafka/consumer.rb

```diff
@@ -93,16 +100,28 @@ def initialize(cluster:, logger:, instrumenter:, group:, fetcher:, offset_manage
     #   checkpoint.
     # @param max_bytes_per_partition [Integer] the maximum amount of data fetched
     #   from a single partition at a time.
+    # @param refresh_topic_interval [Integer] interval of refreshing the topic list.
+    #   Only work when topic_or_regex is a regex
+    #   If it is 0, the topic list won't be refreshed (default)
+    #   If it is n (n > 0), the topic list will be refreshed every n seconds
     # @return [nil]
-    def subscribe(topic_or_regex, default_offset: nil, start_from_beginning: true, max_bytes_per_partition: 1048576)
+    def subscribe(topic_or_regex, default_offset: nil, start_from_beginning: true, max_bytes_per_partition: 1048576, refresh_topic_interval: 0)
       default_offset ||= start_from_beginning ? :earliest : :latest
 
-      if topic_or_regex.is_a?(Regexp)
-        cluster_topics.select { |topic| topic =~ topic_or_regex }.each do |topic|
-          subscribe_to_topic(topic, default_offset, start_from_beginning, max_bytes_per_partition)
-        end
-      else
+      unless topic_or_regex.is_a?(Regexp)
         subscribe_to_topic(topic_or_regex, default_offset, start_from_beginning, max_bytes_per_partition)
+        return
+      end
+
+      subscribe_to_regex(topic_or_regex, default_offset, start_from_beginning, max_bytes_per_partition)
+      if refresh_topic_interval > 0
+        @thread ||= Thread.new do
```

Comment:
Spawning a thread here is error-prone, and hard to control. And like @dasch said, if the user calls this method multiple times, it may spawns multiple threads doing the same thing. In this context, the consumer should only be aware of the new refereshed set of topics before fetching a new fetch. You can choose another approach:
- At the beginning of method `fetch_batches`, check for topic cache expiration. If expired, fetch new list of topics from the cluster. If the topic stay the same, continue.
- If the list of topic changes:
  - Subscribe the new topic to fetcher
  - Subcribe the new topic to group manager
  - Subscribe the new topic to offset manager
  - Mark cluster information as stale, just to be sure
  - Force join group